Libraries/Opc.Ua.Server/Subscription/Subscription.cs

```diff
@@ -847,18 +855,24 @@ private NotificationMessage InnerPublish(
                 // check for monitored items that are ready to publish.
                 LinkedListNode<IMonitoredItem> current = m_itemsToPublish.First;
 
+                //the two lines below fail CTT, but should in theory be able to avoid discarding notifications in a scenario with big durable queues
+                //uint maxNoficationsPerMonitoredItem = (uint)(m_maxNotificationsPerPublish == 0 ? uint.MaxValue : (m_maxNotificationsPerPublish * (ulong)m_maxMessageCount / (ulong)m_itemsToPublish.Count));
+                //maxNoficationsPerMonitoredItem = maxNoficationsPerMonitoredItem == 0 ? 1 : maxNoficationsPerMonitoredItem;
+
+                uint maxNoficationsPerMonitoredItem = m_maxNotificationsPerPublish == 0 ? uint.MaxValue : m_maxNotificationsPerPublish * 3;
+
```

Comment:
This is not sufficient to trigger the moreNotifications output value in ervery edge case.

Limit queue size dynamically depening on no of monitored items?

Unsent messages are dropped in line 969, why don´t we just stop the message creation once there are enough? This way we discard all oldest messages. How do we handle this in a durable scenario. Do we really need a durable messageQueue?

Reply:
@mregen the publishing operation needs further consideration. To trigger the moreNotifications bit in a scenario with few monitored items we need to make shure more thant one notification is generated (m_maxNotificationsPerPublish  +1) however these could come from a single MI. However to avoid the discarding of notifications we need to avoid that more than m_maxMessageCount  notifications are generated per Publish, which depends also on the no. of monitored items. Many server regulate their max queue sizes depending on the no. of monitored items. Maybe we need to calculate the max queue size per (durable) MI depending on the selected Notification Count  (m_maxNotificationsPerPublish * m_maxMessageCount)?

Reply:
the current logic is the following: all MI Queues are completeley emptied every publish cycle and a max of 100 messages (default) can be sent per publish cycle as this is the default max message queue size.

If we put a limit to the published values per MI we need to change the IsReadyToPublish logic of the MI, to only go to false if the publish exhausted the queue.

If we limit the creation of MIs this could change the publishing logic since it continues first with the items being not published in the last publish cylce (whose values would have been lost otherwise).

Reply:
for results of this discussion see #2763